### DISCUSSION TITLE
How to use State or Extension with cf workers?

### DISCUSSION THREAD

**Discussion by makorne** [tokio-rs/axum#1586] [Q&A]
RouteContext<()> is required  for using  KV Stores, Durable Objects and so on in a handler.
I tried State and Extension without success.

```
// #[derive(Clone)]   //RouteContext dont have Clone :(
struct AppState {
    ctx: RouteContext<()>  //Container for a route's parsed parameters, data, and environment bindings from the Runtime (such as KV Stores, Durable Objects, Variables, and Secrets).
}

pub async fn handler(
    request: HttpRequest,
    ctx: Extension<RouteContext<()>>,
) {}

#[event(fetch, respond_with_errors, http)]
pub async fn main(
    req: worker::HttpRequest,
    env: worker::Env,
    _ctx: worker::Context,
) -> Result<axum::response::Response> {

    let mut app=
        axum::Router::new().route("/", axum::routing::get(handler))
            .with_state(AppState {ctx: env.ctx});

    app.call(req)
        .await
        .map_err(|e| worker::Error::from(e.to_string()))
}
```

```
axum::Router::new().route("/", axum::routing::get(handler)).with_state(AppState {ctx: env.ctx});
    |                                        ------------------ ^^^^^^^ the trait `Handler<_, _, _>` is not implemented for fn item `fn(http::Request<worker::ByteStream>, Extension<worker::RouteContext<()>>) -> impl Future<Output = ()> {handler}
```

How to use RouteContext in axum handlers with cf workers?

**Comment by davidpdrsn** [comment]
Try asking the people who maintain that crate. I'm not familiar with it so don't know.

**Comment by makorne** [comment]
Thank you for your great support and your excellent crates!
I often admire your awesome projects like paintings and really think you are one of modern geniuses.
  
And if even you have no idea for this I think there is even smaller chance that they know how to do this.
I and many other users have unanswered simple bugs so I even think may be they don't test their crates at all.

Anyway thank you again for your excellent work for all the world!
